source/slang/slang-check-conversion.cpp

```diff
@@ -423,53 +464,211 @@ namespace Slang
+                        {
+                            coercedArgs.add(coercedArg);
+                        }
                     }
                 }
-
-                // We will go through the fields in order and try to match them
-                // up with initializer arguments.
-                //
-                for(auto fieldDeclRef : getMembersOfType<VarDecl>(m_astBuilder, toStructDeclRef, MemberFilterStyle::Instance))
+            }  
+            else
+            {
+                auto toTypeDeclRef = toDeclRefType->getDeclRef();
+                if (auto toStructDeclRef = toTypeDeclRef.as<StructDecl>())
                 {
-                    Expr* coercedArg = nullptr;
-                    bool argResult = _readValueFromInitializerList(
-                        getType(m_astBuilder, fieldDeclRef),
-                        outToExpr ? &coercedArg : nullptr,
-                        fromInitializerListExpr,
-                        ioArgIndex);
+                    // We will try to coerce the initializer list into a struct following these steps:
+                    // 1. Ensure our `StructDecl` has declaratations created for all auto-generated constructors
+                    // 2. We have a `{}` (0 arguments), try to coerce
+                    // 3. We have a `{arg1, arg2...}`, try to coerce
 
-                    // No point in trying further if any argument fails
-                    if(!argResult)
-                        return false;
+                    auto toStructDecl = toStructDeclRef.getDecl();
+                    // 1. Ensure our `StructDecl` has declaratations created for all auto-generated constructors
+                    ensureDecl(toStructDecl, DeclCheckState::DefaultConstructorReadyForUse);
 
-                    if( coercedArg )
+                    // 2. We have a `{}` (0 arguments), try to coerce
+                    if (argCount == 0)
                     {
-                        coercedArgs.add(coercedArg);
+                        if (outToExpr)
+                        {
+                            *outToExpr = constructZeroInitListFunc(this, toStructDecl, toType, ConstructZeroInitListOptions::CheckToAvoidRecursion);
+                            (*outToExpr)->loc = fromInitializerListExpr->loc;
+                        }
+                        return true;
+                    }
+
+                    // 3. We have a `{arg1, arg2...}`, try to coerce
+                    //
+                    // For this situation we have more steps:
+                    // a. Find constructor candidate for our particular init-list
+                    // b. Collect and coerce init-list arguments into constructor parameters
+                    // c. Create InvokeExpr for our valid ConstructorDecl 
+                    List<ConstructorDecl*> ctorList = _getCtorList(this->getASTBuilder(), this, toStructDecl, nullptr);
+                    bool allowCStyleInitList = checkIfCStyleStruct(this, toStructDecl);
```

Comment:
since `ctorList` is already known here, we should have a variant of `checkIfCStyleStreuct` using the ctorList instead of reconstructing them.

Reply:
addressed here:
https://github.com/shader-slang/slang/pull/5107/commits/54014949789e700a589f5eea098443f3e2bc337c#diff-dad759cd0132d623bee9253550a0bce736d087bdc3157dd1e11e414cb5f6bc48R226